cunumeric/eager.py

```diff
@@ -431,6 +456,57 @@ def diag(self, rhs, extract, k, stacklevel):
             self.array[:] = np.diag(rhs.array, k)
             self.runtime.profile_callsite(stacklevel + 1, False)
 
+    def diag_helper(
+        self,
+        rhs,
+        diag_size,
+        offset,
+        naxes,
+        extract,
+        stacklevel=0,
+        callsite=None,
+    ):
+        if self.shadow:
+            rhs = self.runtime.to_eager_array(rhs, stacklevel=(stacklevel + 1))
+        elif self.deferred is None:
+            self.check_eager_args((stacklevel + 1), rhs)
+        if self.deferred is not None:
+            self.deferred.diag_helper(
+                rhs,
+                diag_size,
+                offset,
+                naxes,
+                extract,
+                stacklevel=(stacklevel + 1),
+            )
+        else:
+            # this should be called only if number of axes <=2
+            if naxes == 2:
+                ndims = rhs.array.ndim
+                self.array[:] = np.diagonal(
+                    rhs.array, offset=offset, axis1=ndims - 2, axis2=ndims - 1
+                )
+            elif naxes < 2:
+                self.array[:] = np.diag(rhs.array, offset)
+            else:
+                ndims = rhs.array.ndim
+                axes = tuple(ax for ax in range(ndims - naxes, ndims))
+                self.array = diagonal_reference(rhs.array, axes)
+            self.runtime.profile_callsite(stacklevel + 1, False)
+
+    def diagonal(self, rhs, offset, axis1, axis2, extract, stacklevel):
+        if self.shadow:
+            rhs = self.runtime.to_eager_array(rhs, stacklevel=(stacklevel + 1))
+        elif self.deferred is None:
+            self.check_eager_args((stacklevel + 1), rhs)
+        if self.deferred is not None:
+            self.deferred.diagonal(
+                rhs, offset, axis1, axis2, extract, stacklevel=(stacklevel + 1)
+            )
+        else:
+            self.array[:] = np.diagonal(rhs.array, offset, axis1, axis2)
```

Comment:
Need to handle case for `extract == True` here

Reply:
for the `diagonal`, `extract` should always be False. Do you mean I should add the check for this?

Reply:
Doesn't matter, since we should remove this function anyway